File: Python/Scripts/ex6_1_1.py
# exercise 6.1.1

from pylab import *
from scipy.io import loadmat
from sklearn import cross_validation, tree

# Load Matlab data file and extract variables of interest
mat_data = loadmat('../Data/wine2.mat')
X = np.matrix(mat_data['X'])
y = np.matrix(mat_data['y'], dtype=int)
attributeNames = [name[0] for name in mat_data['attributeNames'][0]]
classNames = [name[0][0] for name in mat_data['classNames']]
N, M = X.shape
C = len(classNames)

# Tree complexity parameter - constraint on maximum depth
tc = np.arange(2, 21, 1)

# Simple holdout-set crossvalidation
test_proportion = 0.5
X_train, X_test, y_train, y_test = cross_validation.train_test_split(X,y,test_size=test_proportion)

# Initialize variables
Error_train = np.empty((len(tc),1))
Error_test = np.empty((len(tc),1))

for i, t in enumerate(tc):
    # Fit decision tree classifier, Gini split criterion, different pruning levels
    dtc = tree.DecisionTreeClassifier(criterion='gini', max_depth=t)
    dtc = dtc.fit(X_train,y_train.ravel())

    # Evaluate classifier's misclassification rate over train/test data
    y_est_test = dtc.predict(X_test)
    y_est_train = dtc.predict(X_train)
    misclass_rate_test = sum(np.abs(np.mat(y_est_test).T - y_test)) / float(len(y_est_test))
    misclass_rate_train = sum(np.abs(np.mat(y_est_train).T - y_train)) / float(len(y_est_train))
    Error_test[i], Error_train[i] = misclass_rate_test, misclass_rate_train

f = figure(); f.hold(True)
plot(tc, Error_train)
plot(tc, Error_test)
xlabel('Model complexity (max tree depth)')
ylabel('Error (misclassification rate)')
legend(['Error_train','Error_test'])

show()

File: Matlab/Scripts/ex6_1_1.m
% exercise 6.1.1

% Load data
cdir = fileparts(mfilename('fullpath')); 
load(fullfile(cdir,'../Data/wine2'))

% Create holdout crossvalidation partition
CV = cvpartition(classNames(y+1), 'Holdout', .5);

% Pruning levels
prune = 0:10;

% Variable for classification error count
Error_train = nan(1,length(prune));
Error_test = nan(1,length(prune));

% Extract training and test set
X_train = X(CV.training, :);
y_train = y(CV.training);
X_test = X(CV.test, :);
y_test = y(CV.test);

% Fit classification tree to training set
T = classregtree(X_train, classNames(y_train+1), ...
    'method', 'classification', ...
    'splitcriterion', 'gdi', ...
    'categorical', [], ...
    'names', attributeNames, ...
    'prune', 'on', ...
    'minparent', 10);

% Compute classification error
for n = 1:length(prune) % For each pruning level
    Error_train(n) = sum(~strcmp(classNames(y_train+1), eval(T, X_train, prune(n))));
    Error_test(n) = sum(~strcmp(classNames(y_test+1), eval(T, X_test, prune(n))));
end

% Plot classification error
mfig('Wine decision tree: Holdout crossvalidation'); clf; hold all;
plot(prune, Error_train/CV.TrainSize);
plot(prune, Error_test/CV.TestSize);
xlabel('Pruning level');
ylabel('Classification error');
legend('Training error', 'Test error');